prow/plugins/invalidcommitmsg/invalidcommitmsg.go

```diff
@@ -0,0 +1,175 @@
+
+%s
+</details>
+`
+)
+
+var (
+	closeIssueRegex = regexp.MustCompile(`((?i)(clos(?:e[sd]?))|(fix(?:(es|ed)?))|(resolv(?:e[sd]?)))[\s:]+(\w+/\w+)?#(\d+)`)
+	atMentionRegex  = regexp.MustCompile(`\s([@][\w_-]+)`)
+)
+
+func init() {
+	plugins.RegisterPullRequestHandler(pluginName, handlePullRequest, helpProvider)
+}
+
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	// Only the Description field is specified because this plugin is not triggered with commands and is not configurable.
+	return &pluginhelp.PluginHelp{
+			Description: "The invalidcommitmsg plugin applies the '" + invalidCommitMsgLabel + "' label to pull requests whose commit messages contain @ mentions or keywords which can automatically close issues.",
+		},
+		nil
+}
+
+type githubClient interface {
+	AddLabel(owner, repo string, number int, label string) error
+	RemoveLabel(owner, repo string, number int, label string) error
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	CreateComment(owner, repo string, number int, comment string) error
+	ListPRCommits(org, repo string, number int) ([]github.RepositoryCommit, error)
+}
+
+type commentPruner interface {
+	PruneComments(shouldPrune func(github.IssueComment) bool)
+}
+
+func handlePullRequest(pc plugins.Agent, pr github.PullRequestEvent) error {
+	cp, err := pc.CommentPruner()
+	if err != nil {
+		return err
+	}
+	return handle(pc.GitHubClient, pc.Logger, pr, cp)
+}
+
+func handle(gc githubClient, log *logrus.Entry, pr github.PullRequestEvent, cp commentPruner) error {
+	// Only consider actions indicating that the code diffs may have changed.
+	if !hasPRChanged(pr) {
+		return nil
+	}
+
+	shouldComment := pr.Action == github.PullRequestActionSynchronize ||
+		pr.Action == github.PullRequestActionOpened
+
+	var (
+		org    = pr.Repo.Owner.Login
+		repo   = pr.Repo.Name
+		number = pr.Number
+	)
+
+	labels, err := gc.GetIssueLabels(org, repo, number)
+	if err != nil {
+		return err
+	}
+	hasInvalidCommitMsgLabel := github.HasLabel(invalidCommitMsgLabel, labels)
+
+	allCommits, err := gc.ListPRCommits(org, repo, number)
+	if err != nil {
+		return fmt.Errorf("error listing commits for pull request: %v", err)
+	}
+	log.Debugf("Found %d commits in PR", len(allCommits))
+
+	var invalidCommits []github.GitCommit
+	for _, commit := range allCommits {
+		if closeIssueRegex.MatchString(commit.Commit.Message) || atMentionRegex.MatchString(commit.Commit.Message) {
+			c := commit.Commit
+			c.SHA = commit.SHA
+			invalidCommits = append(invalidCommits, c)
+		}
+	}
+
+	// if we have the label but all commits are valid,
+	// remove the label and prune comments
+	if hasInvalidCommitMsgLabel && len(invalidCommits) == 0 {
+		if err := gc.RemoveLabel(org, repo, number, invalidCommitMsgLabel); err != nil {
+			log.WithError(err).Errorf("GitHub failed to remove the following label: %s", invalidCommitMsgLabel)
+		}
+		cp.PruneComments(shouldPrune(log))
+		return nil
```

Comment:
nit: we `return nil` here, but we don't return in later `if` blocks. Is there a particular reason why? It seems like we could let this fall-through, or otherwise add `return nil` elsewhere too? 😄

Reply:
Removed the `nil`. :+1: